### DISCUSSION TITLE
how to implement switch users with supabase auth

### DISCUSSION THREAD

**Discussion by champec** [supabase/supabase#12163] [Questions] [answered]
I'm building an application for a pharmacy that requires different dispensers to log into the same computer app but have a quick way to switch user account like the feature you have on Instagram or YouTube or even Google... allows you to switch account without logging you out. Is this possible with supabase auth? From what I gather once you log in with another account it logs you out the previous session

**Comment by GaryAustin1** [comment, marked as answer]
Not addressing the security of that approach or how to actual implement. But, you can have multiple supabase clients, each with their own session local storage key on a single device/tab in a browser environment.    So multiple users could be signed in and you just pick which client to use for each of them in your app.  I don't know how this carries over if you have server side stuff also involved.

**Reply by champec** [reply]
unfortunately, that doesn't work, supabase only allows one logged-in user per project - no matter how many clients you initiallise because all client use the same auth table - I would have to have multiple projects. Also you can't log in with session data, it has to be a password. The only alternative would be to store passwords in DB. This is driving me crazy

**Reply by GaryAustin1** [reply]
If you do something like:

```
const options = {
  auth: { storageKey: 'key1' }
}
const supabase1 = createClient("https://xyzcompany.supabase.co", "public-anon-key", options)
signin with email/password.
```

Then do a new client with key2
const supabase2 = createClient....
signin with email/password

I think that would work (I've had two clients, but not signed in users on both).
The two clients should not interact, there is no info passed thru REST about the device they are on

Also you can have a user logged in on 10 different devices and only if they signOut do the sessions on all devices get cleared.  This would mean you can't sign the  user out on one device if you want them to log in on another device.

I'll do a test tomorrow on this.

**Reply by GaryAustin1** [reply]
I am able to create two clients and signin two different users with email/password.  Both have sessions separate in local storage and getSession works on both, and database is accessible with either.

![image](https://user-images.githubusercontent.com/54564956/217959287-fc6e42b8-63a3-44c8-aa23-6c82eb4e5a01.png)

```
let supabases = []
supabases[1] = supabase.createClient(SUPABASE_URL, SUPABASE_KEY,{
    auth:{storageKey: 's1'}
})
supabases[2] = supabase.createClient(SUPABASE_URL, SUPABASE_KEY,{
    auth:{storageKey: 's2'}
})
const ses1 = await supabases[1].auth.getSession()
const ses2 = await supabases[2].auth.getSession()
console.log('user1',ses1)
console.log('user2',ses2)
```
Console logs show both users session and user variables.
elsewhere
```
const result = await supabases[1].auth.signInWithPassword({
            email: 'e1@mysite.com',
            password:'aaaaaaaa',
         })
const result2 = await supabases[2].auth.signInWithPassword({
            email: 'e2@mysite.com',
            password:'bbbbbbbb',
        })
```

I believe they will also refresh, when time comes.  If they don't I'll follow up.

Otherwise seems to work for at least email/password signin.   Not sure how easy redirects will be to deal with though so that might take some creativity if using oauth or during the initial signup of a user, but that might be done with a 3rd client and then transferred somehow or require a signin after the signup.   Or use OTP signup/magic link possibly.

**Reply by champec** [reply]
Hi, thanks for your help,  I actuslly didnt try with  storage key options, but I will give that a whirl. Hopefully it works, il give that a try now and let you know

**Reply by champec** [reply]
Hello Gary,

thanks for your help, it seems to work when I store the session in different storage keys. The only issue is it is not dynamic, I have to predetermine a fixed number of clients in order to gain access to all the methods. However it isnt a terrible trade off so I will go that route and have a set maximum number of users

**Reply by GaryAustin1** [reply]
Why a fixed number?   Part of why a tested with an array was so you add users with a signin process and remove with some type of signout process.

**Reply by champec** [reply]
Yeah, actually I suppose I could work out some type of logic to implement that, like a get session function that calls for a new client when they sign in.  However I have implemented a process that stores the access key and refresh token in local storage when they sign in, and then I get the session with getSession auth method. Seems to work at the moment, if I have any problems or issues with this then I will try using your variable client technique.
